src/misc/ethernet_slirp.cpp

```diff
@@ -0,0 +1,411 @@
+	inet_pton(AF_INET, section->Get_string("ipv4_netmask"), &config.vnetmask);
+	inet_pton(AF_INET, section->Get_string("ipv4_host"), &config.vhost);
+	inet_pton(AF_INET, section->Get_string("ipv4_nameserver"),
+	          &config.vnameserver);
+	inet_pton(AF_INET, section->Get_string("ipv4_dhcp_start"),
+	          &config.vdhcp_start);
+	/* IPv6 code is left here as reference but disabled as no DOS-era
+	 * software supports it and might get confused by it */
+	config.in6_enabled = 0;
+	inet_pton(AF_INET6, "fec0::", &config.vprefix_addr6);
+	config.vprefix_len = 64;
+	inet_pton(AF_INET6, "fec0::2", &config.vhost6);
+	inet_pton(AF_INET6, "fec0::3", &config.vnameserver6);
+	/* DHCPv4, BOOTP, TFTP */
+	config.vhostname = "DOSBox-X";
+	config.vdnssearch = NULL;
+	config.vdomainname = NULL;
+	config.tftp_server_name = NULL;
+	config.tftp_path = NULL;
+	config.bootfile = NULL;
+
+	slirp = slirp_new(&config, &slirp_callbacks, this);
+	if (slirp) {
+		LOG_MSG("SLIRP: Successfully initialized");
+		return true;
+	} else {
+		LOG_MSG("SLIRP: Failed to initialize");
+		return false;
+	}
+}
+
+void SlirpEthernetConnection::SendPacket(const uint8_t *packet, int len)
+{
+	slirp_input(slirp, packet, len);
+}
+
+void SlirpEthernetConnection::GetPackets(std::function<void(const uint8_t *, int)> callback)
+{
+	get_packet_callback = callback;
+	uint32_t timeout_ms = 0;
+	PollsClear();
+	PollsAddRegistered();
+	slirp_pollfds_fill(slirp, &timeout_ms, slirp_add_poll, this);
+	bool poll_failed = !PollsPoll(timeout_ms);
+	slirp_pollfds_poll(slirp, poll_failed, slirp_get_revents, this);
+	TimersRun();
+}
+
+void SlirpEthernetConnection::ReceivePacket(const uint8_t *packet, int len)
+{
+	get_packet_callback(packet, len);
+}
+
+struct slirp_timer *SlirpEthernetConnection::TimerNew(SlirpTimerCb cb, void *cb_opaque)
+{
+	struct slirp_timer *timer = new struct slirp_timer;
+	timer->expires = 0;
+	timer->cb = cb;
+	timer->cb_opaque = cb_opaque;
+	timers.push_back(timer);
+	return timer;
+}
+
+void SlirpEthernetConnection::TimerFree(struct slirp_timer *timer)
+{
+	std::remove(timers.begin(), timers.end(), timer);
+	delete timer;
+}
+
+void SlirpEthernetConnection::TimerMod(struct slirp_timer *timer, int64_t expire_time)
+{
+	/* expire_time is in milliseconds despite slirp wanting a nanosecond
+	 * clock */
+	timer->expires = expire_time * 1e6;
+}
+
+void SlirpEthernetConnection::TimersRun()
+{
+	int64_t now = slirp_clock_get_ns(NULL);
+	std::for_each(timers.begin(), timers.end(), [now](struct slirp_timer *&timer) {
+		if (timer->expires && timer->expires < now) {
+			timer->expires = 0;
+			timer->cb(timer->cb_opaque);
+		}
+	});
+}
+
+void SlirpEthernetConnection::TimersClear()
+{
+	std::for_each(timers.begin(), timers.end(),
+	              [](struct slirp_timer *&timer) { delete timer; });
+	timers.clear();
+}
+
+void SlirpEthernetConnection::PollRegister(int fd)
+{
+#ifdef WIN32
+	/* BUG: Skip this entirely on Win32 as libslirp gives us invalid fds. */
+	return;
+#endif
```

Comment:
I haven't reported this upstream since I can't log in to their website

Reply:
Not a problem. I suspect it might be happening because Windows is not really POSIX compliant, so glib on Windows might do something differently than WINAPI here.

Reply:
I'm fairly sure it's a slirp issue, not a glib issue.